katsdpimager/frontend.py

```diff
@@ -34,20 +36,41 @@ def preprocess_visibilities(dataset, args, start_channel, stop_channel,
     else:
         collector = preprocess.VisibilityCollectorMem(
             image_parameters, grid_parameters, args.vis_block)
-    try:
+
+    def reap():
+        nonlocal add_future
+        if add_future is not None:
+            with profile('concurrent.futures.result'):
+                bar_progress = add_future.result()
+            bar.goto(bar_progress)
+            add_future = None
+
+    with contextlib.ExitStack() as exit_stack:
+        exit_stack.callback(collector.close)
+        # We overload data loading with preprocessing by using a separate
+        # thread for preprocessing. To limit memory usage, we only allow
+        # one outstanding chunk at a time.
+        executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
```

Comment:
Would it make a difference to performance if `max_workers` was 2 or 3?

Reply:
No, because I wait for each job to finish before submitting the next one (they're not safe to run concurrently because they'd try to append data to the HDF5 file concurrently; also it bounds memory usage).

Reply:
Ah right, the parallel HDF5 write :-) I was wondering if there was more than just memory bounds involved.